

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Read plugin API &mdash; Mitty 1.1.1dev documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Mitty 1.1.1dev documentation" href="../index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="../index.html" class="fa fa-home"> Mitty</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../introduction.html#installation">1.1. Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../introduction.html#trivia">1.2. Trivia</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">2. Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#creating-genomes">2.1. Creating genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#taking-reads">2.2. Taking reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#testing-alignment-accuracy-of-bwa-mem">2.3. Testing alignment accuracy of BWA MEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#taking-reads-just-from-regions-around-variants">2.4. Taking reads just from regions around variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#analysing-variant-callers">2.5. Analysing variant callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../additional_utils.html">3. Additional Utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../additional_utils.html#plot-gc-bias-plot-gc-bias-of-a-bam-file">3.1. <cite>plot_gc_bias</cite>: Plot GC bias of a BAM file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../theory_and_algorithms.html">4. Theory and algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../theory_and_algorithms.html#population-simulations">4.1. Population simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../models.html">5. Stock Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../models.html#variant-models">5.1. Variant models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../models.html#population-models">5.2. Population models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../models.html#read-models">5.3. Read models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ramblings.html">6. Ramblings AKA. Diary of a Developer</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../ramblings.html#simulating-a-million-genomes-in-python">6.1. Simulating a million Genomes in Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ramblings.html#report-generation">6.2. Report generation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">Mitty</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
    <li>Read plugin API</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="../_sources/dev/reads.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="read-plugin-api">
<h1>Read plugin API<a class="headerlink" href="#read-plugin-api" title="Permalink to this headline">¶</a></h1>
<p>Read Models need to supply two methods:</p>
<p>and</p>
<ol class="arabic simple">
<li>Read plugins have to do more work, because reads and read file format (eg pacbio/sergentis) can be very custom</li>
</ol>
<p>1. We chose coverage per block, rather than reads, because if we did reads sequentially from start to finish of sequence
then we would have a bias in the fastq file. By doing repeated blocks of reads over the whole sequence we ensure reads are
distributed from start to finish at any part of the fastq</p>
<p>For benchmarking, the only thing required from read files is a tagging, done in the qname field, of POS and CIGAR for
that read. Since all reads from the same template need to have the same id if we have M reads as part of the template
we simply chain them all together with a separator:</p>
<div class="highlight-python"><div class="highlight"><pre>tN|POS:CIGAR:D|POS:CIGAR:D|...
</pre></div>
</div>
<p>Here N is the template serial id
| separates fields
POS is the position of the first fragment and so on
: separates values in a field
CIGAR is the CIGAR string
D is the character &#8216;&gt;&#8217; to indicate forward string and &#8216;&lt;&#8217; to indicate reverse string</p>
<div class="section" id="algorithms-for-computing-correct-cigar-and-pos-values-for-simulated-reads">
<h2>Algorithms for computing correct CIGAR and POS values for simulated reads<a class="headerlink" href="#algorithms-for-computing-correct-cigar-and-pos-values-for-simulated-reads" title="Permalink to this headline">¶</a></h2>
<p>TODO: Think of a better algorithm that doesn&#8217;t require so much space and is only invoked when we have variants in the
window</p>
<p>One big goal of Mitty is to serve up realistic test data for bioinformatics algorithms, from aligners to variant
callers. Testing whether a variant caller is correctly working on the simulated data is relatively easy: we simply
compare the variant caller&#8217;s VCF file with the answer book VCF generated by Mitty. It is, however, slightly more involved
to deduce if an aligner is correctly aligning the simulated reads, and to diagnose how the performance of an aligner is
affecting the accuracy of a variant caller. To this end Mitty has a system to compute the correct read position and
CIGAR for each simulated read. This information is stored in the read&#8217;s qname string so that it is easily accessible
to diagnostic programs.</p>
<p>In order to generate reads based on a given VCF file and a reference sequence we go through a two step process.
We first generate the mutated sequence (<cite>mut_seq</cite>) along with some other information that encodes the difference between
each base in the <cite>mut_seq</cite> and corresponding positions on the <cite>ref_seq</cite>. We then generate reads from the <cite>mut_seq</cite> using
the sidecar information to compute the correct POS values and CIGAR strings for the reads.</p>
<p>The algorithm is best introduced through a series of examples. In the examples the reference sequence is labelled <cite>R</cite> and
the mutated sequence is labeled <cite>M</cite>. The information for setting the POS and CIGAR for the read is taken from an
array <cite>pos</cite> that accompanies <cite>M</cite></p>
<div class="section" id="generating-pos">
<h3>Generating <cite>pos</cite><a class="headerlink" href="#generating-pos" title="Permalink to this headline">¶</a></h3>
<p>Let <code class="docutils literal"><span class="pre">R</span> <span class="pre">=</span> <span class="pre">ACTGACTG</span></code></p>
<p>Consider a single base insertion at position 1:</p>
<div class="highlight-python"><div class="highlight"><pre>POS REF ALT
1   A   AT

     1 2345678
R    A CTGACTG
M    ATCTGACTG
pos  1223456789
</pre></div>
</div>
<p>Consider a multiple base insertion at position 1:</p>
<div class="highlight-python"><div class="highlight"><pre>POS REF ALT
1   A   ATT

     1  2345678
R    A  CTGACTG
M    ATTCTGACTG
pos  12223456789
</pre></div>
</div>
<p>Consider a multiple base insertion at last position:</p>
<div class="highlight-python"><div class="highlight"><pre>POS REF ALT
8   G   GTT

     12345678
R    ACTGACTG
M    ACTGACTGTT
pos  12345678999
</pre></div>
</div>
<p>Consider a multiple base deletion:</p>
<div class="highlight-python"><div class="highlight"><pre>POS REF ALT
2   CTG  C

     12345678
R    ACTGACTG
M    AC  ACTG
pos  12  56789
</pre></div>
</div>
<p>Consider a SNP, an insertion and a deletion:</p>
<div class="highlight-python"><div class="highlight"><pre>POS REF ALT
2   C   T
4   G   GTT
6   CTG C

     1234  5678
R    ACTG  ACTG
M    ATTGTTAC
pos  123455569
</pre></div>
</div>
<p><cite>pos</cite> is generated by copying over the index from <cite>R</cite>. When we encounter an insertion we copy over the index of the next
reference base as many times as there is an insertion. Deletions are simply skipped. For the purposes of computing <cite>pos</cite>
we also add an imaginary base position at the end of the reference sequence (9 in this case)</p>
</div>
<div class="section" id="generating-cigars-and-pos-for-reads-from-pos">
<h3>Generating CIGARS and POS for reads from <cite>pos</cite><a class="headerlink" href="#generating-cigars-and-pos-for-reads-from-pos" title="Permalink to this headline">¶</a></h3>
<p>Consider our last example and some reads from <cite>M</cite>:</p>
<div class="highlight-python"><div class="highlight"><pre>     1234  5678
R    ACTG  ACTG
M    ATTGTTAC
pos  123455569
     ++++---------&gt; POS = 1 (The first pos value we encounter)
                    CIGAR = 4M  (2-1=1 -&gt; 1M
                                 3-2=1 -&gt; 1M
                                 4-3=1 -&gt; 1M
                                 5-4=1 -&gt; 1M)

M    ATTGTTAC
pos  123455569
      ++++--------&gt; POS = 2 (The first pos value we encounter)
                    CIGAR = 3M1I  (3-2=1 -&gt; 1M
                                   4-3=1 -&gt; 1M
                                   5-4=1 -&gt; 1M
                                   5-5=0 -&gt; 1I)

M    ATTGTTAC
pos  123455569
       ++++-------&gt; POS = 3
                    CIGAR = 2M2I  (4-3=1 -&gt; 1M
                                   5-4=1 -&gt; 1M
                                   5-5=0 -&gt; 1I
                                   5-5=0 -&gt; 1I)

(A basic aligner would actually soft-clip these last two reads giving us 3M1S and 2M2S)

M    ATTGTTAC
pos  123455569
         ++++-----&gt; POS = 5
                    CIGAR = 2I2M  (5-5=0 -&gt; 1I
                                   5-5=0 -&gt; 1I
                                   6-5=1 -&gt; 1M
                                   9-6=3 -&gt; 1M + 2D) The D only comes into play if our read crosses the deletion
</pre></div>
</div>
<p>To see how a deletion affects our POS and CIGAR consider another previous example:</p>
<div class="highlight-python"><div class="highlight"><pre>POS REF ALT
2   CTG  C

     12345678
R    ACTGACTG
M    AC  ACTG
pos  12  56789
     ++  ++-------&gt; POS = 1
                    CIGAR = 2M2D2M  (2-1=1 -&gt; 1M
                                     5-2=3 -&gt; 1M + 2D The 2D comes into play because the read crosses the boundary
                                     6-5=1 -&gt; 1M
                                     7-6=1 -&gt; 1M)
</pre></div>
</div>
<p>Example of an unmapped read:</p>
<div class="highlight-python"><div class="highlight"><pre>POS REF ALT
2   C  CAATTGG

     12      345678
R    AC      TGACTG
M    ACAATTGGTGACTG
pos  123333333456789
       ++++-------&gt; POS = 3
                    CIGAR = 4I  (3-3=0 -&gt; 1I
                                 3-3=0 -&gt; 1I
                                 3-3=0 -&gt; 1I
                                 3-3=0 -&gt; 1I)
For a read to be mapped, there has to be at least one M. Since there are no Ms we discard the POS and CIGAR as this
is an unmapped read
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">reads.py</span></code> generates simulated reads from <code class="docutils literal"><span class="pre">mut_seq</span></code> based on the read model. Using the <cite>pos</cite> arrays it
also generates appropriate alignment information (POS and CIGAR) that is stored in the qname string.
(Note that while the BAM specs do not place a limit on the length of the qname string both Tablet and IGV expect a
string with length &lt; 255 characters. It is possible that the qname will exceed this and you won&#8217;t be able to open a
set of simulated reads using tools that arbitrarily limit the qname). If no <cite>pos</cite> file is supplied <cite>reads.py</cite> assumes
we are taking reads from a reference sequence and the POS values are actual positions of the reads and all the cigars
are of the form <cite>100M</cite> (For e.g. 100 base reads).</p>
<p>Computing POS: For every read, the POS value is simply the index from <cite>pos</cite> corresponding to the first base of the read
EXCEPT for unmapped reads.</p>
<p>Computing the CIGAR:</p>
<ol class="arabic simple">
<li>Initialize the base counter to <cite>None</cite>, set mapped flag to <cite>False</cite></li>
<li>Step through the each base of the read and look at the difference in <cite>pos</cite> values <cite>dp</cite></li>
<li>If <cite>dp==1</cite>, if the counter is any thing other than <cite>M</cite>, flush it. Set or increment counter as <cite>M</cite>. Set mapped flag to <cite>True</cite></li>
<li>If <cite>dp==0</cite>, if the counter is other than <cite>I</cite>, flush it. Set or increment counter as <cite>I</cite></li>
<li>If <cite>dp&gt;1</cite>, if the counter is other than <cite>M</cite>, flush it. Set and flush counter as <cite>M</cite>, set counter as <cite>D</cite> to be dp-1</li>
<li>Continue from 2 until done.</li>
<li>Flush any counter other than <cite>D</cite></li>
<li>If the mapped flag is <cite>False</cite> reset POS and CIGAR - this is an unmapped read.</li>
</ol>
<p>You can &#8220;read along&#8221; to these examples by running <cite>python reads.py test -v</cite> and seeing how different functions in
<cite>reads.py</cite> implement these algorithms</p>
</div>
<div class="section" id="jit-expansion-of-variant-sequences">
<h3>JIT expansion of variant sequences<a class="headerlink" href="#jit-expansion-of-variant-sequences" title="Permalink to this headline">¶</a></h3>
<p>TODO</p>
</div>
</div>
</div>


          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2015,Seven Bridges Genomics
(kaushik.ghose@sbgenomics.com).
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.1.1dev',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>