

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Writing variant plugins &mdash; Mitty 1.1.1dev documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Mitty 1.1.1dev documentation" href="../index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="../index.html" class="fa fa-home"> Mitty</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../introduction.html#installation">1.1. Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../introduction.html#trivia">1.2. Trivia</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">2. Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#creating-genomes">2.1. Creating genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#taking-reads">2.2. Taking reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#testing-alignment-accuracy-of-bwa-mem">2.3. Testing alignment accuracy of BWA MEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#taking-reads-just-from-regions-around-variants">2.4. Taking reads just from regions around variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#analysing-variant-callers">2.5. Analysing variant callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../additional_utils.html">3. Additional Utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../additional_utils.html#plot-gc-bias-plot-gc-bias-of-a-bam-file">3.1. <cite>plot_gc_bias</cite>: Plot GC bias of a BAM file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../theory_and_algorithms.html">4. Theory and algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../theory_and_algorithms.html#population-simulations">4.1. Population simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../models.html">5. Stock Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../models.html#variant-models">5.1. Variant models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../models.html#population-models">5.2. Population models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../models.html#read-models">5.3. Read models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ramblings.html">6. Ramblings AKA. Diary of a Developer</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../ramblings.html#simulating-a-million-genomes-in-python">6.1. Simulating a million Genomes in Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ramblings.html#report-generation">6.2. Report generation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">Mitty</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
    <li>Writing variant plugins</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="../_sources/dev/variant_plugins.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="writing-variant-plugins">
<span id="how-to-write-variant-plugins"></span><h1>Writing variant plugins<a class="headerlink" href="#writing-variant-plugins" title="Permalink to this headline">Â¶</a></h1>
<p>A few example variant generators are found under <code class="docutils literal"><span class="pre">plugins/variants</span></code>. Plugins are registered using the
<code class="docutils literal"><span class="pre">mitty.plugins.variants</span></code> entry point. Examples of this can be seen in the <code class="docutils literal"><span class="pre">setup.py</span></code> script for the supplied
variant generators. We will inspect relevant parts of the supplied <code class="docutils literal"><span class="pre">snp</span></code> plugin to see how we can write plugins.</p>
<p>The plugin module needs to expose one generator function accepting at least two parameters:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">variant_generator</span><span class="p">(</span><span class="n">ref</span><span class="o">=</span><span class="p">{},</span> <span class="n">master_seed</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">ref</span></code> is the reference chromosome as a <code class="xref py py-class docutils literal"><span class="pre">mitty.lib.genome.FastaGenome</span></code> object. <code class="docutils literal"><span class="pre">master_seed</span></code> is an integer
<code class="docutils literal"><span class="pre">0</span> <span class="pre">&lt;</span> <span class="pre">master_seed</span> <span class="pre">&lt;</span> <span class="pre">mitty.lib.SEED_MAX</span></code>. The seed is generated by the calling program (such as <code class="docutils literal"><span class="pre">denovo</span></code>) and is
meant to be used by any random number generators run by the model. The seed is generated from the master_seed passed
via the input parameter file.</p>
<p>The function should return a generator. Each run of the iterator should return a dictionary whose keys are the
chromosome number the variants are for and the value should be the list of proposed variants
(<code class="xref py py-class docutils literal"><span class="pre">mitty.lib.variation.Variation</span></code>), like so:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">chromosome</span><span class="p">:</span>
  <span class="n">ref_chrom</span> <span class="o">=</span> <span class="n">ref</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
  <span class="n">snp_locs</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">place_poisson</span><span class="p">(</span><span class="n">base_loc_rng</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_chrom</span><span class="p">))</span>
  <span class="n">het_type</span> <span class="o">=</span> <span class="n">util</span><span class="o">.</span><span class="n">zygosity</span><span class="p">(</span><span class="n">snp_locs</span><span class="o">.</span><span class="n">size</span><span class="p">,</span> <span class="n">phet</span><span class="p">,</span> <span class="n">het_rng</span><span class="p">,</span> <span class="n">copy_rng</span><span class="p">)</span>
  <span class="n">base_subs</span> <span class="o">=</span> <span class="n">base_sub_rng</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">snp_locs</span><span class="o">.</span><span class="n">size</span><span class="p">)</span>

  <span class="k">yield</span> <span class="p">{</span><span class="n">chrom</span><span class="p">:</span> <span class="p">[</span><span class="n">new_variation</span><span class="p">(</span><span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">pos</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span> <span class="n">ref_chrom</span><span class="p">[</span><span class="n">pos</span><span class="p">],</span> <span class="n">base_sub_dict</span><span class="p">[</span><span class="n">ref_chrom</span><span class="p">[</span><span class="n">pos</span><span class="p">]][</span><span class="n">bs</span><span class="p">],</span> <span class="n">zygosity</span><span class="p">)</span>
                 <span class="k">for</span> <span class="n">pos</span><span class="p">,</span> <span class="n">bs</span><span class="p">,</span> <span class="n">zygosity</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">snp_locs</span><span class="p">,</span> <span class="n">base_subs</span><span class="p">,</span> <span class="n">het_type</span><span class="p">)</span> <span class="k">if</span> <span class="n">ref_chrom</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">]}</span>
  <span class="c"># +1 because VCF files are 1 indexed</span>
  <span class="c">#alts will be 0 if ref is not one of ACTG</span>
</pre></div>
</div>
<p>In this case we loop over each chromosome, sending out a list of variants each time we loop. It is perfectly legal to
return variants on multiple chromosomes as well.</p>
<p><strong>It is important that the returned variants are sorted in ascending order. If this violated then subsequent processing
steps will fail.</strong></p>
<p>In addition to the generator function the module should also expose:</p>
<ul>
<li><p class="first">An example parameter string. This is used for the help as well as for the integration test that the test suite runs on all discovered plugins:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">__example_param_text</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">{</span>
<span class="s">  &quot;chromosome&quot;: [4],      # List of chromosomes to apply the variant to</span>
<span class="s">  &quot;p&quot;: 0.01,              # probability that the SNP will happen at any given base</span>
<span class="s">  &quot;phet&quot;: 0.5,            # probability that the variant will be heterozygous</span>
<span class="s">}</span>
<span class="s">&quot;&quot;&quot;</span>
</pre></div>
</div>
</li>
<li><p class="first">A description text that is printed when the model is described:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">_description</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">This is the stock SNP plugin. A typical parameter set resembles</span>
<span class="s">&quot;&quot;&quot;</span> <span class="o">+</span> <span class="n">__example_param_text</span>
</pre></div>
</div>
</li>
<li><p class="first">A dictionary that is passed to the integration test as an example parameter set. I usually use the same string as for the description so I don&#8217;t have to repeat myself:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">_example_params</span> <span class="o">=</span> <span class="nb">eval</span><span class="p">(</span><span class="n">__example_param_text</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
<p>Typically, you will be distributing your plugin as a separate Python package. You need to register the package as a
Mitty plugin by indicating an entry point in your <code class="docutils literal"><span class="pre">setup.py</span></code> script.</p>
<p>Assuming that your module is named <code class="docutils literal"><span class="pre">snp_plugin.py</span></code> and the directory structure of your code is:</p>
<div class="highlight-python"><div class="highlight"><pre>mitty_snp
     |--&gt; snp_plugin.py
setup.py
</pre></div>
</div>
<p>Your <code class="docutils literal"><span class="pre">setup.py</span></code> script would indicate the entry point as:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">entry_points</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;mitty.plugins.variants&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s">&#39;mysnp = mitty_snp.snp_plugin&#39;</span><span class="p">]}</span>
</pre></div>
</div>
<p>Users would call this model by the name <code class="docutils literal"><span class="pre">mysnp</span></code> in their parameter files. Python currently allows you to add duplicate
entry point names. If names are duplicated, though the plugin will be found on the list of available plugins, in the
simulations, the plugin with the same name that was installed first will be the one that is used.</p>
<p>A complete <code class="docutils literal"><span class="pre">setup.py</span></code> would look like:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">setuptools</span> <span class="kn">import</span> <span class="n">setup</span><span class="p">,</span> <span class="n">find_packages</span>

<span class="n">setup</span><span class="p">(</span>
    <span class="n">name</span><span class="o">=</span><span class="s">&#39;mitty_plugin_snp&#39;</span><span class="p">,</span>
    <span class="n">version</span><span class="o">=</span><span class="s">&#39;1.0.0&#39;</span><span class="p">,</span>
    <span class="n">description</span><span class="o">=</span><span class="s">&#39;An example set of plugins for Mitty&#39;</span><span class="p">,</span>
    <span class="n">author</span><span class="o">=</span><span class="s">&#39;Seven Bridges Genomics&#39;</span><span class="p">,</span>
    <span class="n">author_email</span><span class="o">=</span><span class="s">&#39;kaushik.ghose@sbgenomics.com&#39;</span><span class="p">,</span>
    <span class="n">packages</span><span class="o">=</span><span class="n">find_packages</span><span class="p">(</span><span class="n">include</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;mitty_snp&#39;</span><span class="p">]),</span>
    <span class="n">entry_points</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;mitty.plugins.variants&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s">&#39;mysnp = mitty_snp.snp_plugin&#39;</span><span class="p">]}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>


          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2015,Seven Bridges Genomics
(kaushik.ghose@sbgenomics.com).
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.1.1dev',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>