

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Genome simulation &mdash; Mitty 1.1.1dev documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Mitty 1.1.1dev documentation" href="../index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="../index.html" class="fa fa-home"> Mitty</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../introduction.html#installation">1.1. Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../introduction.html#trivia">1.2. Trivia</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">2. Quick-start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html#creating-genomes">2.1. Creating genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html#taking-reads">2.2. Taking reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html#testing-alignment-accuracy-of-bwa-mem">2.3. Testing alignment accuracy of BWA MEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html#taking-reads-just-from-regions-around-variants">2.4. Taking reads just from regions around variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html#analysing-variant-callers">2.5. Analysing variant callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../additional_utils.html">3. Additional Utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../additional_utils.html#plot-gc-bias-plot-gc-bias-of-a-bam-file">3.1. <cite>plot_gc_bias</cite>: Plot GC bias of a BAM file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../theory_and_algorithms.html">4. Theory and algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../theory_and_algorithms.html#population-simulations">4.1. Population simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../models.html">5. Stock Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../models.html#variant-models">5.1. Variant models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../models.html#population-models">5.2. Population models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../models.html#read-models">5.3. Read models</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">Mitty</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
    <li>Genome simulation</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="../_sources/dev/variants.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="genome-simulation">
<span id="variation-struct"></span><h1>Genome simulation<a class="headerlink" href="#genome-simulation" title="Permalink to this headline">¶</a></h1>
<p>There are two kinds of genome simulations:</p>
<blockquote>
<div><ul class="simple">
<li>site frequency spectrum sampled population (Sampled)</li>
<li>detailed forward simulation population (Forward)</li>
</ul>
</div></blockquote>
<p>The <cite>Sampled</cite> simulation is generated by creating a large list of potential variants and then selecting, with replacement,
from this list, to create a chromosome. This is done twice per sample to create a diploid genome. Each variant has a
probability (<span class="math">\(p_i\)</span>) of appearing in a sample. The distribution of <span class="math">\(p_i\)</span> is called the site frequency spectrum.
The probability for a heterozygous variant is <span class="math">\(p_i^2\)</span></p>
<p>A <cite>Forward</cite> simulation starts with a population generated by Sampling and is then carried forward, generation by
generation using models of sexual reproduction, including recombination of chromosomes.</p>
<div class="section" id="population-variant-master-list">
<h2>Population Variant master list<a class="headerlink" href="#population-variant-master-list" title="Permalink to this headline">¶</a></h2>
<p>The main simulation asks each variant plugin for a list of candidate variants that are then stored in a sorted master
list of variants. Each variant has an associated probability of it appearing in any given chromosome. The distribution of the
probabilities follow the shape given by the site frequency spectrum models.</p>
<p>For the Single and Sampled simulations we select variants (with replacement, for each chromosome) from this list to create
new chromosomes. For Single and Sampled simulations we select variants from the whole list.</p>
<p>For Forward simulations, the founder population is created using the Sampled simulation algorithm but by restricting
the variant selection to those having a certain minimum probability value. These initial variants are then marked off the
list. Subsequent generations add the remaining lower probability variants without replacement. The simulation will
terminate when we run out of variants or when the required number of generations have been run.</p>
</div>
</div>
<div class="section" id="data-structures">
<h1>Data structures<a class="headerlink" href="#data-structures" title="Permalink to this headline">¶</a></h1>
<p>Internally, a variant is represented by a light-weight container called <code class="xref py py-class docutils literal"><span class="pre">mitty.lib.variation.Variant</span></code>. All the
variants are represented in a master list sorted by pos value.</p>
<p>A chromosome is represented as a linked list of a <code class="xref py py-class docutils literal"><span class="pre">mitty.lib.variation.GTVariant</span></code>. When we do operations on
a chromosome of a sample we always go through each variant sequentially, so this is a good data structure for our use
case.</p>
</div>
<div class="section" id="storage">
<h1>Storage<a class="headerlink" href="#storage" title="Permalink to this headline">¶</a></h1>
<p>The results of the simulation are stored in a sqlite database. We can write VCFs from this database and the read generator
can read genomes directly off this database to generate reads.</p>
</div>
<div class="section" id="detailed-algorithms">
<h1>Detailed Algorithms<a class="headerlink" href="#detailed-algorithms" title="Permalink to this headline">¶</a></h1>
<div class="section" id="site-frequency-spectrum">
<h2>Site frequency spectrum<a class="headerlink" href="#site-frequency-spectrum" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="selecting-variants-from-the-master-list">
<h2>Selecting variants from the master list<a class="headerlink" href="#selecting-variants-from-the-master-list" title="Permalink to this headline">¶</a></h2>
<p>UPDATE!!!!!</p>
<p>We have V variants in our master list, and each variant has an associated probability p_k of being included in a
sample (k=0, 1, .... V-1). The distribution of p_k follows the site frequency spectrum, as discussed above.</p>
<p>In principle we would generate N random numbers from a uniform distribution between 0.0 and 1.0 and use that to
decide whether each individual variant is selected for a sample or not. In practice N can be 10 million or more and so
we look to reduce our computational costs.</p>
<p>We quantize the x-axis (variant probability) of the site frequency spectrum into M levels (p_i, i=0,1,2 ... M-1) each
with v_i variants. V=sum(v_i) i=0,1,2 ... M-1. Since each variant in a given probability bucket has the same p_i value
we can use a poisson model to pick variants from each bucket without having to go through each individual variant.
This dramatically reduces our selection cost and makes it proportional to the number of variants/sample individual, rather
than proportional to the number of variants in the master list.</p>
<p>As a side note, sum(p_i * v_i) = E(variants in a sample).</p>
<p>We create the master list from all the variants from all the variant generators and then sort it by position. We then
create M arrays with <cite>indexes</cite> of the variants that fall into a given probability bin. Each array is sorted by
probability value, and then all the arrays are re-balanced so that the distribution now follows the ideal site frequency
distribution. (We sort the arrays first because we want to maintain the rank-ordering of the points even after re-balancing)</p>
<p>We then run M separate poisson searches to pick indexes from each bucket. We combine and sort the indexes and then
create the candidate chromosome by ensuring none of the variants clash with each other, and merging any homozygous
variants as needed.</p>
<p>TODO: Detailed description of Forward sim.</p>
</div>
</div>
<div class="section" id="variant-plugins">
<h1>Variant plugins<a class="headerlink" href="#variant-plugins" title="Permalink to this headline">¶</a></h1>
<p>The main simulation engine passes the following parameters to a variant plugin:</p>
<div class="highlight-python"><div class="highlight"><pre>Chromosome  - indicates which chromosome is being simulated
start, stop - indicates the coordinate range, along this chromosome, that is being simulated
ref         - the reference sequence of this chromosome as a string
p, f        - the site frequency spectrum the simulation is using.
</pre></div>
</div>
<p>The variant plugin returns a list of variations. Each returned variation should have an associated probability value
to indicate the relative rarity of the mutation. An extremely rare mutation should have a
value close to 0.0 while an extremely common one should have a value of 0.1 or so. These values are re-normalized by the
simulation so that the overall site frequency spectrum of all generated variants follows the required site frequency spectrum.</p>
<p>In addition to the above parameters, a variant plugin may take a custom list of parameters (such as mutation rate) for
a one time initialization.</p>
<div class="section" id="future-directions">
<h2>Future directions<a class="headerlink" href="#future-directions" title="Permalink to this headline">¶</a></h2>
<p>Can generate linkage by picking neighboring variants, rather than arbitrary variants</p>
</div>
</div>


          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014,Seven Bridges Genomics
(kaushik.ghose@sbgenomics.com).
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.1.1dev',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>