Mitty is a collection of programs that enable us to generate simulated genomic data to test our algorithms.

Mitty generates simulated data by starting from a reference genome and then adding controlled mutations to it to create
a variant "sample" genome. Mitty can also simulate populations of genomes derived via sexual reproduction. Mitty can
then generate reads from the chromosomes of the mutated genome(s). 

At each stage of this process we know the origins of the data and so we know

- The correct alignment of each read
- The correct variants in the "sample" that these reads came from.
- For population simulations, the lineage and relatedness of the samples

**The data generated by Mitty can, therefore, be used to test the correctness of aligners and variant callers under
different conditions and data set characteristics.**

Installation
============
    
    pip install 









The following command should be sufficient on a Mac to get the `develop` branch of Mitty running on your machine

    pip install git+https://gitlab.sbgenomics.com:9443/graphgenome/mitty.git@develop#egg=mitty


Linux specific preamble
-----------------------
As needed (i.e. if you haven't ever done it)

    sudo apt-get install python-pip
    sudo apt-get install git
    sudo apt-get install python-dev    

The easy way
------------
Add the following

    [global]
    extra-index-url = https://pip.sbgenomics.com/packages

to your `~/.pip/pip.conf` file (creating if needed) and then `pip install` from the repository

    sudo pip install git+https://gitlab.sbgenomics.com:9443/graphgenome/mitty.git@develop

* Enter your credentials as needed
* You should be using the develop branch. Periodically master will catchup, but develop has the latest stuff.


Run the tests
-------------

    pip install nose  #If needed
    nosetests -v


The long way
------------

    pip install cython
    pip install tox   # If you want to run tests using tox
    pip install nose  # If you want to run the tests to make sure everything is ok
    
Get and install the code
------------------------
    
    git clone git@gitlab.sbgenomics.com:graphgenome/mitty.git
    cd mitty
    git checkout develop  # Usually, as ab SBG-er you should be using this
    pip install .
    
          
Updating
--------
If you already have an older version of mitty you should 

    git pull
    pip install -e    


Developing
----------

To develop on Mitty, simply clone the repository, and from the project root run ::

    pip install -e .


Overview
========

Mitty, at its heart, is a set of command line programs that are called with various parameters to generate different 
simulated data. Mitty has, additionally, been wrapped for the Seven Bridges Genomics computation platform (IGOR) and
can be used from that platform to generate simulated data.

The important components of Mitty are described as::

    Reference Genome  + denovo.py      -->  .vcf file
        mutations of various kinds representing a single sample
  

    Reference Genome  + population.py  -->  lot of .vcf files
        a founder population of genomes
      

    Reference Genome + founder population  + generations.py  -->  lot of .vcf files, lineage files
        multiple generations created by sexual reproduction
      
      
    .vcf file + vcf2reads.py           -->   .fastq file
        perfect and corrupted reads from "sample"
    
    .fastq file + reads2bam.py         -->   .bam file
        perfect alignment of "sample" reads to original reference
    
    
    .bam file + checkbam.py            -->  diagnostic files
        Allows us to check the quality of our alignments

Please see the documentation for a quickstart and detailed manual