

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Mitty components &mdash; Mitty 1.1.1dev documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Mitty 1.1.1dev documentation" href="index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="index.html" class="fa fa-home"> Mitty</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="introduction.html">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#installation">1.1. Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#trivia">1.2. Trivia</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">2. Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#creating-genomes">2.1. Creating genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#taking-reads">2.2. Taking reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#testing-alignment-accuracy-of-bwa-mem">2.3. Testing alignment accuracy of BWA MEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#taking-reads-just-from-regions-around-variants">2.4. Taking reads just from regions around variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#analysing-variant-callers">2.5. Analysing variant callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="additional_utils.html">3. Additional Utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="additional_utils.html#plot-gc-bias-plot-gc-bias-of-a-bam-file">3.1. <cite>plot_gc_bias</cite>: Plot GC bias of a BAM file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="theory_and_algorithms.html">4. Theory and algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="theory_and_algorithms.html#population-simulations">4.1. Population simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Stock Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="models.html#variant-models">5.1. Variant models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#population-models">5.2. Population models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#read-models">5.3. Read models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ramblings.html">6. Ramblings AKA. Diary of a Developer</a><ul>
<li class="toctree-l2"><a class="reference internal" href="ramblings.html#simulating-a-million-genomes-in-python">6.1. Simulating a million Genomes in Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="ramblings.html#report-generation">6.2. Report generation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Mitty</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Mitty components</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/components.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="mitty-components">
<h1>Mitty components<a class="headerlink" href="#mitty-components" title="Permalink to this headline">¶</a></h1>
<div class="section" id="create-single-or-multiple-genomes">
<h2>Create single or multiple genomes<a class="headerlink" href="#create-single-or-multiple-genomes" title="Permalink to this headline">¶</a></h2>
<p>(For a detailed description of the algorithms/structures please see <code class="xref doc docutils literal"><span class="pre">genomes</span></code>)</p>
<span class="target" id="module-mitty.genomes"></span><p>genomes (1.9.0.dev0): Generate simulated genomes from a simulation parameter file.</p>
<p>Commandline:</p>
<div class="highlight-python"><div class="highlight"><pre>Usage:
  genomes generate &lt;pfile&gt;  [-v|-V] [-p]
  genomes dryrun &lt;pfile&gt;
  genomes inspect &lt;dbfile&gt;
  genomes inspect sfs &lt;chrom&gt; &lt;dbfile&gt;
  genomes write vcf &lt;dbfile&gt; &lt;out_prefix&gt;  [--sample_name=SN] [-v|-V]
  genomes explain (parameters|(variant-model|spectrum-model|population-model) (all|&lt;model_name&gt;))
  genomes list (variant-model|spectrum-model|population-model)

Options:
  generate                Create a database of genomes by running the models specified
  &lt;pfile&gt;                 Name for parameter file
  dryrun                  Don&#39;t run simulation, but print out useful info about it
  inspect                 Give summary information about a genome database
  &lt;dbfile&gt;                Name of genome database file
  sfs                     Print the site frequency spectrum
  &lt;chrom&gt;                 Which chromosome&#39;s site frequncy spectrum to print
  write                   Write out genome data from the database file in vcf format
  vcf                     Write out all genomes in one multi-sample vcf file
  &lt;out_prefix&gt;            Written files will have this prefix
  --sample_name=SN        Name of sample. Leave out to write master list
  explain                 Explain the parameters/variant model/population model
  all                     Iterate over and explain all models
  &lt;model_name&gt;            Explain specific model
  list                    List the models
  -v                      Dump log messages
  -V                      Dump detailed log messages
  -p                      Show progress barParameter file example::

{
  # Path notes: an absolute path is left as is. A relative path is taken relative to the parameter file location
  &quot;files&quot;: {
    &quot;reference_dir&quot;: &quot;/Users/kghose/Data/hg38/&quot;,  # If reference is chr1.fa, chr2.fa ... in this directory
    &quot;reference_file&quot;: &quot;/Users/kghose/Data/hg38/hg38.fa.gz&quot;,  # If reference is a single gz fasta file
    &quot;dbfile&quot;: &quot;Out/test.db&quot;  # Output database file
  },
  &quot;rng&quot;: {
    &quot;master_seed&quot;: 1
  },
  &quot;population_model&quot;: {
    &quot;standard&quot;: {            # Name of sample chooser (population) model.
      &quot;sample_size&quot;: 1,      # population model parameters
    }
  },
  &quot;site_model&quot;: {
      &quot;double_exp&quot;: {    # Name of model that handles the site frequency spectrum
        &quot;k1&quot;: 0.1,       # spectrum model parameters
        &quot;k2&quot;: 2.0,
        &quot;p0&quot;: 0.001,
        &quot;p1&quot;: 0.2,
        &quot;bin_cnt&quot;: 30
      }
    }
  &quot;chromosomes&quot;: [1, 2]  # Chromosomes to apply the models to
  &quot;variant_models&quot;: [    # The list of variant models should come under this key
    {
      &quot;snp&quot;: {           # name of the model.
        &quot;p&quot;: 0.01        # Parameters required by the model
      }
    },
    {                    # We can chain as many models as we wish. We can also repeat models.
      &quot;delete&quot; : {
        &quot;p&quot;: 0.01
      }
    }
  ]
}
</pre></div>
</div>
<div class="section" id="create-a-population-of-genomes-population-py">
<h3>Create a population of genomes: <code class="docutils literal"><span class="pre">population.py</span></code><a class="headerlink" href="#create-a-population-of-genomes-population-py" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="create-multiple-generations-via-sexual-reproduction-generations-py">
<h3>Create multiple generations via sexual reproduction : <code class="docutils literal"><span class="pre">generations.py</span></code><a class="headerlink" href="#create-multiple-generations-via-sexual-reproduction-generations-py" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="generate-reads">
<h2>Generate reads<a class="headerlink" href="#generate-reads" title="Permalink to this headline">¶</a></h2>
<p>(For a detailed description of algorithms please see <code class="xref doc docutils literal"><span class="pre">reads</span></code>)</p>
<div class="section" id="generate-reads-from-a-genome-vcf2reads-py">
<h3>Generate reads from a genome: <code class="docutils literal"><span class="pre">vcf2reads.py</span></code><a class="headerlink" href="#generate-reads-from-a-genome-vcf2reads-py" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="diagnostic-tools">
<h2>Diagnostic tools<a class="headerlink" href="#diagnostic-tools" title="Permalink to this headline">¶</a></h2>
<div class="section" id="generate-a-perfect-alignment-reads2bam">
<h3>Generate a perfect alignment: <code class="docutils literal"><span class="pre">reads2bam</span></code><a class="headerlink" href="#generate-a-perfect-alignment-reads2bam" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="analyze-an-alignment-checkbam">
<h3>Analyze an alignment: <code class="docutils literal"><span class="pre">checkbam</span></code><a class="headerlink" href="#analyze-an-alignment-checkbam" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2015,Seven Bridges Genomics
(kaushik.ghose@sbgenomics.com).
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.1dev',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>