

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>1. Overview &mdash; Mitty 1.1.1dev documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Mitty 1.1.1dev documentation" href="index.html"/>
        <link rel="next" title="2. Tutorial" href="tutorial.html"/>
        <link rel="prev" title="Mitty documentation" href="index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="index.html" class="fa fa-home"> Mitty</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installation">1.1. Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#trivia">1.2. Trivia</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">2. Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#creating-genomes">2.1. Creating genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#taking-reads">2.2. Taking reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#testing-alignment-accuracy-of-bwa-mem">2.3. Testing alignment accuracy of BWA MEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#taking-reads-just-from-regions-around-variants">2.4. Taking reads just from regions around variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#analysing-variant-callers">2.5. Analysing variant callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="additional_utils.html">3. Additional Utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="additional_utils.html#plot-gc-bias-plot-gc-bias-of-a-bam-file">3.1. <cite>plot_gc_bias</cite>: Plot GC bias of a BAM file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="theory_and_algorithms.html">4. Theory and algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="theory_and_algorithms.html#population-simulations">4.1. Population simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Stock Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="models.html#variant-models">5.1. Variant models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#population-models">5.2. Population models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#read-models">5.3. Read models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ramblings.html">6. Ramblings AKA. Diary of a Developer</a><ul>
<li class="toctree-l2"><a class="reference internal" href="ramblings.html#simulating-a-million-genomes-in-python">6.1. Simulating a million Genomes in Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="ramblings.html#report-generation">6.2. Report generation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Mitty</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>1. Overview</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/introduction.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="overview">
<h1>1. Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>Mitty is a collection of programs to generate simulated datasets to benchmark aligners and variant callers.</p>
<p>Mitty generates simulated data by starting from a reference genome and then adding controlled mutations to it to create
one or more variant &#8220;sample&#8221; genomes. Mitty can then generate reads from the chromosomes of the mutated genome(s).</p>
<p>At each stage of this process we know the origins of the data and so we know</p>
<ul class="simple">
<li>The correct alignment of each read</li>
<li>The correct variants in the &#8220;sample&#8221; that these reads came from.</li>
</ul>
<p><strong>The data generated by Mitty can, therefore, be used to test the correctness of aligners and variant callers under
different conditions and data set characteristics.</strong></p>
<p>Simulated data generation:</p>
<div class="highlight-python"><div class="highlight"><pre>  fasta-----------------|-------------------------------------------|
(reference)             |                                           |
                  |-----------|      |-----------|             |-----------|
  variant -------&gt;| genome    |-----&gt;| simulated |------------&gt;| read      |
  models          | simulator |      | genomes   |             | simulator |-----&gt; simulated
                  |           |      |-----------|       |----&gt;|           |       fastq
  population ----&gt;|-----------|        |                 |     |-----------|
  model                 |              |                read         |
                        |              |                model        |
                    parameters         |--&gt; simulated             parameters
                                              VCF
</pre></div>
</div>
<p>Algorithm benchmarking:</p>
<div class="highlight-python"><div class="highlight"><pre>                |---------|                |----------|     |--------|     |----------|
simulated       | Aligner |                | BAM      |     | BAM    |     | Analysis |
fastq     -----&gt;| being   |-----&gt; BAM ----&gt;| analyzer |----&gt;| stats  |----&gt;| reports  |
                | tested  |        |       |----------|     |--------|     |----------|
                |---------|        |
                                   |
                                   |    |----------|
                                   |    | Variant  |             |----------|    |----------|
                                   |    | caller   |             | VCF      |    | Analysis |
                                   |---&gt;| being    |---&gt; VCF ---&gt;| analyzer |---&gt;| reports  |
                                        | tested   |             |----------|    |----------|
                                        |----------|                 |
                                                         simulated --|
                                                         VCF
</pre></div>
</div>
<div class="section" id="installation">
<h2>1.1. Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<div class="section" id="virtual-environments">
<h3>1.1.1. Virtual environments<a class="headerlink" href="#virtual-environments" title="Permalink to this headline">¶</a></h3>
<p>While not required, it is strongly recommended that you use <a class="reference external" href="http://docs.python-guide.org/en/latest/dev/virtualenvs/">virtual environments</a> for your Python code. I find
<a class="reference external" href="https://virtualenvwrapper.readthedocs.org/en/latest/">virtualenvwrapper</a> awesome for managing Python virtual environments:</p>
<div class="highlight-python"><div class="highlight"><pre>sudo pip install virtualenvwrapper
source /usr/local/bin/virtualenvwrapper.sh  # You may want to add this to your login script
</pre></div>
</div>
<p>(You should study the details of installing this tool from their excellent documentation)</p>
<p>Then create a new environment for your Mitty install:</p>
<div class="highlight-python"><div class="highlight"><pre>mkvirtualenv mitty-test
</pre></div>
</div>
<p>This will prepare a new environment and activate it. When you are done you can do:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">deactivate</span>
</pre></div>
</div>
<p>And when you want to come back do:</p>
<div class="highlight-python"><div class="highlight"><pre>workon mitty-test
</pre></div>
</div>
</div>
<div class="section" id="mitty-installation">
<h3>1.1.2. Mitty installation<a class="headerlink" href="#mitty-installation" title="Permalink to this headline">¶</a></h3>
<p>On Linux, you may first need to pull some dependencies (as needed) such as</p>
<div class="highlight-python"><div class="highlight"><pre>sudo apt-get install python-pip git python-numpy python-dev cython zlib1g-dev libhdf5-serial-dev
</pre></div>
</div>
<p>You will also need setuptools 11.0.0 or more recent</p>
<div class="highlight-python"><div class="highlight"><pre>pip install -U setuptools
</pre></div>
</div>
<p>The SBG dev pi server contains wheels for most of the packages, including the latest builds of Mitty. You can use the
devpi server (cuts down on compilation time in many cases) by adding the following lines to your ~/.pip/pip.conf file
(you may need to create the file first):</p>
<div class="highlight-python"><div class="highlight"><pre>[global]
extra-index-url = https://pip.sbgenomics.com/packages
</pre></div>
</div>
<p>To get the stable release version of Mitty try</p>
<div class="highlight-python"><div class="highlight"><pre>pip install -U mitty
</pre></div>
</div>
<p>For the dev team, you will want the <cite>develop</cite> branch of Mitty</p>
<div class="highlight-python"><div class="highlight"><pre>pip install -U mitty --pre
</pre></div>
</div>
<p>To get the code directly from gitlab (e.g. because you want a specific commit/tag) use something like</p>
<div class="highlight-python"><div class="highlight"><pre>pip install git+https://gitlab.sbgenomics.com:9443/graphgenome/mitty.git@develop#egg=mitty
</pre></div>
</div>
<p>On mac os x this will pull all packages as needed, including numpy.</p>
<div class="section" id="installing-optional-dependencies">
<h4>1.1.2.1. Installing optional dependencies<a class="headerlink" href="#installing-optional-dependencies" title="Permalink to this headline">¶</a></h4>
<p>The analysis report generators, like <cite>plot_align</cite> may require additional packages, like matplotlib, to run. To install
these too you should add the tag <cite>[mplot]</cite> to the install command:</p>
<div class="highlight-python"><div class="highlight"><pre>pip install git+https://gitlab.sbgenomics.com:9443/graphgenome/mitty.git@develop#egg=mitty[mplot]
</pre></div>
</div>
</div>
</div>
<div class="section" id="run-the-tests">
<h3>1.1.3. Run the tests<a class="headerlink" href="#run-the-tests" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>pip install nose  # If needed
nosetests mitty -v
</pre></div>
</div>
</div>
<div class="section" id="developing">
<h3>1.1.4. Developing<a class="headerlink" href="#developing" title="Permalink to this headline">¶</a></h3>
<p>To develop on Mitty, simply clone the repository, and from the project root run</p>
<div class="highlight-python"><div class="highlight"><pre>pip install -e .
</pre></div>
</div>
<p>or, for the optional, plotting dependencies</p>
<div class="highlight-python"><div class="highlight"><pre>pip install -e .[mplot]
</pre></div>
</div>
<div class="section" id="creating-wheels">
<h4>1.1.4.1. Creating wheels<a class="headerlink" href="#creating-wheels" title="Permalink to this headline">¶</a></h4>
<div class="highlight-python"><div class="highlight"><pre>pip install wheel  # If needed
pip wheel .
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="trivia">
<h2>1.2. Trivia<a class="headerlink" href="#trivia" title="Permalink to this headline">¶</a></h2>
<p>Since you were dying to know: Mitty comes from James Thurber&#8217;s <a class="reference external" href="http://www.newyorker.com/archive/1939/03/18/390318fi_fiction_thurber?currentPage=all">&#8220;The Secret Life of Walter Mitty&#8221;</a> one of my
favourite pieces from one of my favourite authors. Though <a class="reference external" href="http://en.wikipedia.org/wiki/The_Secret_Life_of_Walter_Mitty">Wikipedia</a> has a less favourable interpretation of what
Walter Mitty stands for, I always thought that Thurber was celebrating the dreamer within each us who spices up the banal
parts of life with a little harmless fantasy.</p>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="tutorial.html" class="btn btn-neutral float-right" title="2. Tutorial">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="Mitty documentation"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2015,Seven Bridges Genomics
(kaushik.ghose@sbgenomics.com).
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.1dev',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>