

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>2. Quick-start &mdash; Mitty 1.1.1dev documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Mitty 1.1.1dev documentation" href="index.html"/>
        <link rel="next" title="3. Additional Utilities" href="additional_utils.html"/>
        <link rel="prev" title="1. Overview" href="introduction.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="index.html" class="fa fa-home"> Mitty</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#installation">1.1. Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#trivia">1.2. Trivia</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">2. Quick-start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#creating-genomes">2.1. Creating genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#taking-reads">2.2. Taking reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="#testing-alignment-accuracy-of-bwa-mem">2.3. Testing alignment accuracy of BWA MEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="#taking-reads-just-from-regions-around-variants">2.4. Taking reads just from regions around variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="#analysing-variant-callers">2.5. Analysing variant callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="additional_utils.html">3. Additional Utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="additional_utils.html#plot-gc-bias-plot-gc-bias-of-a-bam-file">3.1. <cite>plot_gc_bias</cite>: Plot GC bias of a BAM file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="theory_and_algorithms.html">4. Theory and algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="theory_and_algorithms.html#population-simulations">4.1. Population simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Stock Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="models.html#variant-models">5.1. Variant models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#population-models">5.2. Population models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#read-models">5.3. Read models</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Mitty</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>2. Quick-start</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/quickstart.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="quick-start">
<h1>2. Quick-start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h1>
<p>This quickstart introduces the most commonly used features of Mitty. For additional commands, utilities and usages
please see <a class="reference internal" href="additional_utils.html#additional-utilities"><span>Additional Utilities</span></a></p>
<p>We will introduce Mitty via a data simulation problem. Let us say our colleague Heng Li has written an aligner
called <code class="docutils literal"><span class="pre">bwa</span></code> and a variant caller called <code class="docutils literal"><span class="pre">samtools</span></code> and we want to figure out how well they do using synthetic data.</p>
<p>Our plan is as follows:</p>
<ul class="simple">
<li>Consider a small reference genome (we will use the Red Alga, <em>Cyanidioschyzon merolae strain 10D</em>, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/assembly/GCF_000091205.1/#/def">record</a>, <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000091205.1_ASM9120v1">ftp</a>)</li>
<li>Sprinkle some variants on the reference genome to create a population of 1000 sample genomes (why not?)</li>
<li>Take 30x Illumina like reads from one of the sample genomes</li>
<li>Use <cite>bwa</cite> to align the reads back to the reference genome</li>
<li>Assess how well <cite>bwa</cite> aligned the reads</li>
</ul>
<div class="section" id="creating-genomes">
<h2>2.1. Creating genomes<a class="headerlink" href="#creating-genomes" title="Permalink to this headline">¶</a></h2>
<p>The example reference genome has 20 chromosomes. We&#8217;d like to add SNPs, insertions and deletions to chromosome 1 and 2.
We will need the <code class="docutils literal"><span class="pre">genomes</span></code> program for this</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes
Generate simulated genomes from a simulation parameter file.

Commandline::

  Usage:
    genomes generate &lt;pfile&gt;  [-v|-V] [-p]
    genomes dryrun &lt;pfile&gt;
    genomes inspect &lt;dbfile&gt;
    genomes inspect sfs &lt;chrom&gt; &lt;dbfile&gt;
    genomes write ((vcf|vcfs)|master) &lt;dbfile&gt; &lt;out_prefix&gt; [&lt;serial&gt;]... [-v|-V]
    genomes explain (parameters|(variantmodel|populationmodel) (all|&lt;model_name&gt;))
    genomes list (variantmodels|populationmodels)

  Options:
    generate                Create a database of genomes by running the models specified
    &lt;pfile&gt;                 Name for parameter file
    dryrun                  Don&#39;t run simulation, but print out useful info about it
    inspect                 Give summary information about a genome database
    &lt;dbfile&gt;                Name of genome database file
    sfs                     Print the site frequency spectrum
    &lt;chrom&gt;                 Which chromosome&#39;s site frequncy spectrum to print
    write                   Write out genome data from the database file in vcf format
    vcf                     Write out all genomes in one multi-sample vcf file
    vcfs                    Write out the genomes in separate single sample vcf files
    master                  Write  a VCF file with all the variants, no sample tag
    &lt;serial&gt;                Serial number of sample. Not needed and Ignored if &#39;master&#39; is used. Leave blank with vcf/vcfs to write all samples
    &lt;out_prefix&gt;            Written files will have this prefix
    explain                 Explain the parameters/variant model/population model
    all                     Iterate over and explain all models
    &lt;model_name&gt;            Explain specific model
    list                    List the models
    -v                      Dump log messages
    -V                      Dump detailed log messages
    -p                      Show progress bar
</pre></div>
</div>
<p>We need the <code class="docutils literal"><span class="pre">genomes</span> <span class="pre">generate</span></code> tool, and we need a refresher on how to write a simulation parameter file</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes explain parameters
Parameter file example::

  {
    # Path notes: an absolute path is left as is. A relative path is taken relative to the parameter file location
    &quot;files&quot;: {
      &quot;reference_dir&quot;: &quot;/Users/kghose/Data/hg38/&quot;,  # If reference is chr1.fa, chr2.fa ... in this directory
      &quot;reference_file&quot;: &quot;/Users/kghose/Data/hg38/hg38.fa.gz&quot;,  # If reference is a single gz fasta file
      &quot;dbfile&quot;: &quot;Out/test.db&quot;  # Output database file
    },
    &quot;rng&quot;: {
      &quot;master_seed&quot;: 1
    },
    &quot;sample_size&quot;: 1,      # How many samples to generate
    &quot;site_model&quot;: {
        &quot;double_exp&quot;: {    # Name of model that handles the site frequency spectrum
          &quot;k1&quot;: 0.1,       # Population model parameters
          &quot;k2&quot;: 2.0,
          &quot;p0&quot;: 0.001,
          &quot;p1&quot;: 0.2,
          &quot;bin_cnt&quot;: 30
        }
      }
    &quot;chromosomes&quot;: [1, 2]  # Chromosomes to apply the models to
    &quot;variant_models&quot;: [    # The list of variant models should come under this key
      {
        &quot;snp&quot;: {           # name of the model.
          &quot;p&quot;: 0.01        # Parameters required by the model
        }
      },
      {                    # We can chain as many models as we wish. We can also repeat models.
        &quot;delete&quot; : {
          &quot;p&quot;: 0.01
        }
      }
    ]
  }
</pre></div>
</div>
<p>What are the variant models available to us?</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes list variantmodels

Available variant models
----------------
- delete (mitty.plugins.variants.delete_plugin)

- insert (mitty.plugins.variants.insert_plugin)

- snp (mitty.plugins.variants.snp_plugin)
</pre></div>
</div>
<p>What kind of parameters do they need?</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes explain variantmodel snp

---- snp (mitty.plugins.variants.snp_plugin) ----

This is the stock SNP plugin. A typical parameter set resembles

{
  &quot;p&quot;: 0.01,              # probability that the SNP will happen at any given base
  &quot;t_mat&quot;: [[ 0.32654629,  0.17292732,  0.24524503,  0.25528135],  # Base transition matrix
            [ 0.3489394,   0.25942695,  0.04942584,  0.3422078],   # Leading diagonal is ignored
            [ 0.28778188,  0.21087004,  0.25963262,  0.24171546],
            [ 0.21644706,  0.20588717,  0.24978216,  0.32788362]],
}

Model defaults: (p=0.01,  t_mat=None)
</pre></div>
</div>
<p>What are the population models available to us?</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes list populationmodels

Available population models
----------------
- double_exp (mitty.plugins.site_frequency.double_exp)
</pre></div>
</div>
<p>An example parameter snippet?</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes explain populationmodel double_exp

---- double_exp (mitty.plugins.site_frequency.double_exp) ----
This stock site frequency plugin employs a double exponential as a simple site frequency model.

f = a1 * exp(-p * k1) + a2 * exp(-p * k2)   (f is normalized such that sum(f) = 1.0)

Example parameters:

{
  &quot;double_exp&quot;: {
    &quot;a1&quot;: 1.0,        # Amplitude of component 1
    &quot;k1&quot;: 20.0,       # decay rate of component 1
    &quot;a2&quot;: 0.1,        # Amplitude of component 2
    &quot;k2&quot;: 5.0,        # Decay rate of component 2
    &quot;p0&quot;: 0.001,      # Lower end of probability range
    &quot;p1&quot;: 0.2,        # Upper end of probability range
    &quot;bin_cnt&quot;: 30     # Number of bins for probability distribution
  }
}

Model defaults: (a1=1.0,  k1=20.0,  a2=0.1,  k2=5.0,  p0=0.001,  p1=0.2,  bin_cnt=21)
</pre></div>
</div>
<p>Let us create a parameter file for the simulations, calling it <code class="docutils literal"><span class="pre">variations.json</span></code>:</p>
<div class="highlight-json"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41</pre></div></td><td class="code"><div class="highlight"><pre><span class="p">{</span>
  <span class="nt">&quot;files&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="nt">&quot;reference_file&quot;</span><span class="p">:</span> <span class="s2">&quot;../data/red_alga.fa.gz&quot;</span><span class="p">,</span>
    <span class="nt">&quot;dbfile&quot;</span><span class="p">:</span> <span class="s2">&quot;Out/red_alga_genomes.db&quot;</span>
  <span class="p">},</span>
  <span class="nt">&quot;rng&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="nt">&quot;master_seed&quot;</span><span class="p">:</span> <span class="mi">12345</span>
  <span class="p">},</span>
  <span class="nt">&quot;sample_size&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
  <span class="nt">&quot;site_model&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="nt">&quot;double_exp&quot;</span><span class="p">:</span> <span class="p">{</span>
      <span class="nt">&quot;a1&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
      <span class="nt">&quot;k1&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
      <span class="nt">&quot;a2&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
      <span class="nt">&quot;k2&quot;</span><span class="p">:</span> <span class="mf">5.0</span><span class="p">,</span>
      <span class="nt">&quot;p0&quot;</span><span class="p">:</span> <span class="mf">0.001</span><span class="p">,</span>
      <span class="nt">&quot;p1&quot;</span><span class="p">:</span> <span class="mf">0.8</span><span class="p">,</span>
      <span class="nt">&quot;bin_cnt&quot;</span><span class="p">:</span> <span class="mi">30</span>
    <span class="p">}</span>
  <span class="p">},</span>
  <span class="nt">&quot;chromosomes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
  <span class="nt">&quot;variant_models&quot;</span><span class="p">:</span> <span class="p">[</span>          
    <span class="p">{</span>
      <span class="nt">&quot;snp&quot;</span><span class="p">:</span> <span class="p">{</span>                 
        <span class="nt">&quot;p&quot;</span><span class="p">:</span> <span class="mf">0.001</span>
      <span class="p">}</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="nt">&quot;delete&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="nt">&quot;p&quot;</span><span class="p">:</span> <span class="mf">0.0001</span><span class="p">,</span>
        <span class="nt">&quot;p_end&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
        <span class="nt">&quot;del_len_min&quot;</span><span class="p">:</span> <span class="mi">2</span>
      <span class="p">}</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="nt">&quot;insert&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="nt">&quot;p&quot;</span><span class="p">:</span> <span class="mf">0.0001</span>
      <span class="p">}</span>
    <span class="p">}</span>
  <span class="p">]</span>
<span class="p">}</span>
</pre></div>
</td></tr></table></div>
<p>Before we run the actual simulation, let&#8217;s make sure the site frequency spectrum looks satisfactory</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes dryrun ../examples/demo/variations.json
Site frequency model
 p                 f
0.001 --------------------------------------------------------------------------------0.351
0.029 ------------------------------------------------0.212
0.056 ------------------------------0.130
0.084 -------------------0.082
0.111 ------------0.054
0.139 --------0.037
0.166 ------0.026
0.194 ----0.019
0.221 ---0.015
0.249 ---0.012
0.277 --0.009
0.304 --0.008
0.332 --0.007
0.359 -0.006
0.387 -0.005
0.414 -0.004
0.442 -0.004
0.469 -0.003
0.497 -0.003
0.524 -0.002
0.552 0.002
0.580 0.002
0.607 0.002
0.635 0.001
0.662 0.001
0.690 0.001
0.717 0.001
0.745 0.001
0.772 0.001
0.800 0.001

1/sum(p_i * f_i) = 13.9

Growth curve:

N	v
1	0.072033
5	0.239221
10	0.352144
50	0.607667
100	0.670136
500	0.787052
1000	0.870873
5000	0.997640
10000	0.999984
50000	1.000000
</pre></div>
</div>
<p>Let&#8217;s run this command and create a database of simulated genomes</p>
<div class="highlight-text"><div class="highlight"><pre>$ mkdir -p ../examples/demo/Out
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ genomes generate ../examples/demo/variations.json -v
WARNING:mitty.genomes:Removed old simulation file
DEBUG:mitty.genomes:Took 7.601594s
DEBUG:mitty.genomes:14742 unique variants, 1862331 variants in samples
</pre></div>
</div>
<p>(<em>Using the -v option will give detailed logger messages, using the -p option will give a progress bar</em>).</p>
<p>Let&#8217;s take a peek at the produced genomes database using the inspect function</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes inspect ../examples/demo/Out/red_alga_genomes.db
../examples/demo/Out/red_alga_genomes.db
	2 chromosomes
	1000 samples
Unique variants in population
	Chrom	Variants
	1	7063
	2	7679
Variants in samples
	Chrom	Avg variants	Std variants	Variants/megabase
	1	888.17   	21.97    	2101.6
	2	973.08   	27.54    	2129.2
</pre></div>
</div>
<p>We can also take a look at the site frequency spectrum in the generated population, for chromosome 1, for example</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes inspect sfs 1 ../examples/demo/Out/red_alga_genomes.db
Site frequency spectrum for chrom 1
0.00 -------------------------------------------------------------------------------- 3977
0.08 -------------------------------------- 1884
0.16 ------------ 576
0.24 ----- 252
0.32 --- 142
0.40 - 64
0.48 - 67
0.56 - 45
0.64 - 29
0.72  19
0.80  8
</pre></div>
</div>
<p>Right now, all our genomes are in a database. The rest of the world works in VCF files, so let&#8217;s write out one of the
samples as a VCF file</p>
<div class="highlight-text"><div class="highlight"><pre>$ genomes write vcf ../examples/demo/Out/red_alga_genomes.db ../examples/demo/Out/red_alga 3
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ head -n 20 ../examples/demo/Out/red_alga_s3.vcf
##fileformat=VCFv4.1
##contig=&lt;ID=NC_010127.1,length=422616,md5=fe4be2f3bc5a7754085ceaa39a5c0414&gt;
##contig=&lt;ID=NC_010128.1,length=457013,md5=99880025dcbcba5dbf72f437092903c3&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	s3
NC_010127.1	29	.	T	G	100	PASS	.	GT	0|1
NC_010127.1	231	.	AC	A	100	PASS	.	GT	1|0
NC_010127.1	909	.	C	T	100	PASS	.	GT	0|1
NC_010127.1	1871	.	G	A	100	PASS	.	GT	1|0
NC_010127.1	2033	.	T	G	100	PASS	.	GT	1|1
NC_010127.1	3853	.	G	A	100	PASS	.	GT	1|1
NC_010127.1	4117	.	TT	T	100	PASS	.	GT	1|0
NC_010127.1	4981	.	T	G	100	PASS	.	GT	1|0
NC_010127.1	5061	.	G	T	100	PASS	.	GT	0|1
NC_010127.1	5440	.	AA	A	100	PASS	.	GT	0|1
NC_010127.1	6038	.	C	T	100	PASS	.	GT	0|1
NC_010127.1	6056	.	A	T	100	PASS	.	GT	1|0
NC_010127.1	6201	.	T	TCTTGC	100	PASS	.	GT	0|1
NC_010127.1	6381	.	T	TCAGGGGGTAGGGGTATCAAAATGCTCATGTAAA	100	PASS	.	GT	0|1
NC_010127.1	6723	.	G	C	100	PASS	.	GT	0|1
</pre></div>
</div>
<p>Things seem to have run satisfactorily. Note that we produce phased VCF files and we also use the notation <cite>1|0</cite> since we
have complete knowledge of the phasing. Now let&#8217;s generate a bag of reads from this genome.</p>
</div>
<div class="section" id="taking-reads">
<h2>2.2. Taking reads<a class="headerlink" href="#taking-reads" title="Permalink to this headline">¶</a></h2>
<p>We will use the <cite>reads</cite> program to generate a fastq file of reads from the selected genome.</p>
<div class="highlight-text"><div class="highlight"><pre>$ reads
This program generates reads from a genome (either a reference genome or a sample genome).
The read characteristics are governed by the chosen read plugin.

Commandline::

  Usage:
    reads generate &lt;pfile&gt;  [-v|-V] [-p]
    reads list
    reads explain (parameters|model (all|&lt;model_name&gt;))

  Options:
    &lt;pfile&gt;         Name for parameter file
    list            List available models
    explain         Explain details about the indicated model
    parameters      Explain parameters
    model           Explain a model
    all             Iterate over all models and explain them
    &lt;model_name&gt;    The model to explain
    -p              Show progress bar
    -v              Dump detailed logger messages
    -V              Dump very detailed logger messages
</pre></div>
</div>
<p>We know how to get help for writing the parameter file</p>
<div class="highlight-text"><div class="highlight"><pre>$ reads explain parameters
Parameter file example::

  {
    &quot;files&quot;: {
      # An absolute path is left as is
      # a relative path is taken relative to the location of the *script*
      &quot;reference_dir&quot;: &quot;/Users/kghose/Data/hg38/&quot;,  # Use this if the reference consists of multiple .fa files in a directory
      &quot;reference_file&quot;: &quot;/Users/kghose/Data/hg38/hg38.fa.gz&quot;,  # Use this if reference is a single multi-fasta file
      &quot;dbfile&quot;: &quot;Out/test.db&quot;  # Genomes database file. Leave out if taking reads from a reference, or from VCF file
      &quot;input_vcf&quot;: &quot;Out/test.vcf&quot;,  # Use this if using a VCF file. Leave out if taking reads from a reference, or from VCF file
      &quot;output_prefix&quot;: &quot;Out/reads&quot;, # Output file name prefix
                                    # the reads will be called reads.fq and reads_c.fq if we call for corrupted reads too
      &quot;interleaved&quot;: true     # Set to false if you need separate files for each mate of a pair
                              # files will then be named reads_1.fq, reads_c_1.fq and reads_2.fq, reads_c_2.fq
                              # If the reads are actually not paired and you set interleaved to false you will get two
                              # files _1 and _2 and all the data will be in _1 only
    },
    &quot;sample&quot; : {
      &quot;gen&quot;: 0,      # Use gen and serial for database
      &quot;serial&quot;: 1,
      &quot;name&quot;: &quot;HN23456&quot; # Use name for VCF file
    }
    &quot;rng&quot;: {
      &quot;master_seed&quot;: 1
    },
    &quot;chromosomes&quot;: [1,2],               # List the chromosomes the reads should be taken from
    &quot;variants_only&quot;: false,             # If true, reads will only come from the vicinity of variants
    &quot;variant_window&quot;: 500,              # Only used if variants_only is true. Reads will be taken from this vicinity
    &quot;corrupt&quot;: true,                    # If true, corrupted reads will also be written
    &quot;coverage&quot;: 10,                     # Coverage
    &quot;coverage_per_block&quot;: 0.01          # For each block of the simulation we generate reads giving this coverage
                                        # In some simulations we will have too many reads to fit in memory and we must
                                        # write them out in blocks of smaller size.
    &quot;read_model&quot;: &quot;simple_illumina&quot;,    # Model specific parameters, need to be under the key &quot;model_params&quot;
    &quot;model_params&quot;: {
      &quot;read_len&quot;: 100,          # length of each read
      &quot;template_len_mean&quot;: 250, # mean length of template
      &quot;template_len_sd&quot;: 30,    # sd of template length
      &quot;max_p_error&quot;: 0.01,      # Maximum error rate at tip of read
      &quot;k&quot;: 20                   # exponent
    }

  }
</pre></div>
</div>
<p>What sort of read models are currently available?</p>
<div class="highlight-text"><div class="highlight"><pre>$ reads list

Available models
----------------
- simple_illumina (mitty.plugins.reads.simple_illumina_plugin)

- simple_sequential (mitty.plugins.reads.simple_sequential_plugin)
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">simple_illumina</span></code> sounds promising, what kind of parameter file does it need?</p>
<div class="highlight-text"><div class="highlight"><pre>$ reads explain model simple_illumina
This is the stock read plugin that approximates Illumina reads

GC bias model based on:

&quot;Summarizing and correcting the GC content bias in high-throughput sequencing&quot;, Benjamini and Speed,
Nucleic Acids Research, 2012
http://nar.oxfordjournals.org/content/early/2012/02/08/nar.gks001.full

1. &quot;GC effect is mostly driven by the GC composition of the full fragment.&quot;
2. &quot;the GC curve is unimodal is key to this analysis. In all data sets shown,
the rate of GC-poor or GC-rich fragments is significantly lower than average, in many cases zero.&quot;

Seven Bridges Genomics
Current contact: kaushik.ghose@sbgenomics.com

Example parameter set:

{
  &#39;read_len&#39;: 100,          # length of each read
  &#39;template_len_mean&#39;: 250, # mean length of template
  &#39;template_len_sd&#39;: 30,    # sd of template length
  &#39;max_p_error&#39;: 0.01,      # Maximum error rate at tip of read
  &#39;k&#39;: 20,                  # exponent
  &#39;gc_bias&#39;: {              # Omit if not modeling GC bias
    &#39;bias_center&#39;: 0.5,     # Center of uni-modal window
    &#39;bias_height&#39;: 1.5,     # Maximum departure from mean coverage (peak of curve)
    &#39;bias_spread&#39;: 0.3      # Width measure of curve
  }
}

Model defaults: (read_len=100,  template_len_mean=250,  template_len_sd=50,  max_p_error=0.01,  k=20,  gc_bias=None)
</pre></div>
</div>
<p>Ok, let&#8217;s put together a parameter file for our experiment, let&#8217;s call it <cite>illumina_reads.json</cite>:</p>
<div class="highlight-json"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26</pre></div></td><td class="code"><div class="highlight"><pre><span class="p">{</span>
  <span class="nt">&quot;files&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="nt">&quot;reference_file&quot;</span><span class="p">:</span> <span class="s2">&quot;../data/red_alga.fa.gz&quot;</span><span class="p">,</span>
    <span class="nt">&quot;output_prefix&quot;</span><span class="p">:</span> <span class="s2">&quot;Out/reads&quot;</span><span class="p">,</span>
    <span class="nt">&quot;dbfile&quot;</span><span class="p">:</span> <span class="s2">&quot;Out/red_alga_genomes.db&quot;</span>
  <span class="p">},</span>
  <span class="nt">&quot;sample&quot;</span> <span class="p">:</span> <span class="p">{</span>
    <span class="nt">&quot;gen&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="nt">&quot;serial&quot;</span><span class="p">:</span> <span class="mi">3</span>
  <span class="p">},</span>
  <span class="nt">&quot;rng&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="nt">&quot;master_seed&quot;</span><span class="p">:</span> <span class="mi">123456</span>
  <span class="p">},</span>
  <span class="nt">&quot;chromosomes&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
  <span class="nt">&quot;read_model&quot;</span><span class="p">:</span> <span class="s2">&quot;simple_illumina&quot;</span><span class="p">,</span>
  <span class="nt">&quot;coverage&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">,</span>
  <span class="nt">&quot;coverage_per_block&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
  <span class="nt">&quot;corrupt&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
  <span class="nt">&quot;model_params&quot;</span><span class="p">:</span> <span class="p">{</span>
    <span class="nt">&quot;read_len&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
    <span class="nt">&quot;template_len_mean&quot;</span><span class="p">:</span> <span class="mi">250</span><span class="p">,</span>
    <span class="nt">&quot;template_len_sd&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
    <span class="nt">&quot;max_p_error&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="nt">&quot;k&quot;</span><span class="p">:</span> <span class="mi">100</span>
  <span class="p">}</span>
<span class="p">}</span>
</pre></div>
</td></tr></table></div>
<p>Let&#8217;s run this command and create some reads from the variant genome</p>
<div class="highlight-text"><div class="highlight"><pre>$ reads generate ../examples/demo/illumina_reads.json -v
DEBUG:mitty.reads:Took 25.402335s to write 280992 templates
</pre></div>
</div>
</div>
<div class="section" id="testing-alignment-accuracy-of-bwa-mem">
<h2>2.3. Testing alignment accuracy of BWA MEM<a class="headerlink" href="#testing-alignment-accuracy-of-bwa-mem" title="Permalink to this headline">¶</a></h2>
<p>First, use BWA-MEM to create an alignment:</p>
<div class="highlight-text"><div class="highlight"><pre>$ bwa index ../examples/data/red_alga.fa.gz
[bwa_index] Pack FASTA... 0.22 sec
...
[main] CMD: bwa index ../examples/data/red_alga.fa.gz
[main] Real time: 13.353 sec; CPU: 13.195 sec
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ bwa mem -t 8 -p ../examples/data/red_alga.fa.gz  ../examples/demo/Out/reads_c.fq &gt; ../examples/demo/Out/temp.sam
[M::main_mem] read 561984 sequences (56198400 bp)...
...
[main] CMD: bwa mem -t 8 -p ../examples/data/red_alga.fa.gz ../examples/demo/Out/reads_c.fq
[main] Real time: 17.182 sec; CPU: 52.714 sec
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ samtools view -Sb  ../examples/demo/Out/temp.sam &gt; ../examples/demo/Out/temp.bam
[samopen] SAM header is present: 20 sequences.
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ samtools sort ../examples/demo/Out/temp.bam  ../examples/demo/Out/reads
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ samtools index ../examples/demo/Out/reads.bam
</pre></div>
</div>
<p>We can use an alignment browser, such as <a class="reference internal" href="#tablet" id="id1">[Tablet]</a>, to see the alignments</p>
<img alt="_static/reads_bam_tablet.png" src="_static/reads_bam_tablet.png" />
<p>(You can see that the coverage is what we dialled in. The one red colored base is a read error - since we are inspecting
the corrupted reads file)</p>
<table class="docutils citation" frame="void" id="tablet" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[Tablet]</a></td><td>Milne I, Stephen G, Bayer M, Cock PJA, Pritchard L, Cardle L, Shaw PD and Marshall D. 2013. Using Tablet for visual exploration of second-generation sequencing data. Briefings in Bioinformatics 14(2), 193-202. (<a class="reference external" href="http://ics.hutton.ac.uk/tablet/">web</a>)</td></tr>
</tbody>
</table>
<div class="section" id="analyse-the-alignment">
<h3>2.3.1. Analyse the alignment<a class="headerlink" href="#analyse-the-alignment" title="Permalink to this headline">¶</a></h3>
<p>We use <code class="docutils literal"><span class="pre">perfectbam</span></code> to both analyze the performance of the aligner and to create a perfectly aligned <code class="docutils literal"><span class="pre">.bam</span></code> file from the
<code class="docutils literal"><span class="pre">.bam</span></code> file output by the aligner.</p>
<div class="highlight-text"><div class="highlight"><pre>$ perfectbam ../examples/demo/Out/reads.bam
</pre></div>
</div>
<p><cite>perfectbam</cite> creates a <cite>.json</cite> file with some summary performance statistics and a <cite>.csv</cite> file with details about the
misaligned reads.</p>
<div class="highlight-text"><div class="highlight"><pre>$ head -n 15 ../examples/demo/Out/reads_summary.json
{
  &quot;read_counts&quot;: {
    &quot;1&quot;: 126204, 
    &quot;3&quot;: 144470, 
    &quot;2&quot;: 137152, 
    &quot;4&quot;: 154164
  }, 
  &quot;incorrectly_aligned_read_counts&quot;: {
    &quot;1&quot;: 3079, 
    &quot;3&quot;: 2635, 
    &quot;2&quot;: 3771, 
    &quot;4&quot;: 2012
  }, 
  &quot;sequence_header&quot;: [
    {
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ head -n 10 ../examples/demo/Out/reads_misaligned.csv
qname	error_type	correct_chrom	correct_pos	correct_cigar	aligned_chrom	aligned_pos	aligned_cigar	mapping_qual	mate_is_unmapped	seq
r268293|4|1|1|513131|100=|0|512973|100=, 3, 4, 513131, 100M, 1, 225, 100M, 0, False, ATGCGACCCTAACTTTCGTAGCGATGCTGCAAACAAAACTCTAGCTTACTTGCACCTGTGCCCCTTACTATCAACCAAGTAATTATTTTTTGCAACTCCA
r268293|4|1|1|513131|100=|0|512973|100=, 3, 4, 512973, 100M, 1, 384, 100M, 0, False, TCTCTTGGTTTTTTTTCCTGGCCTCTCTTTGTCAACGACTGGTTATCTAGCCTTTCTTGTACTCCCGTTCGGGGATATGTCTTTGTGCTGATTTTCATCG
r229601|4|0|0|512789|100=|1|512955|100=, 3, 4, 512955, 100M, 1, 402, 100M, 0, False, TGGCCTCTCTTTGTCAACGACTGGTTATCTAGCCTTTCTTGTACTCCCGTTCGGGGATATGTCTTTGTGCTGATTTTCATCGAGGCGCTACGCATCCTTA
r268288|4|1|1|512873|100=|0|512723|100=, 3, 4, 512873, 100M, 1, 484, 100M, 0, False, AGGCGCTACGCATCCTTACCCATGGGCTACGGCTCGTCACTAATTTTTTATCGATTAGGGTGCTGTGAAGCGATTGGTTTCGGTCTGTGTCTTTTTTCTT
r229601|4|0|0|512789|100=|1|512955|100=, 3, 4, 512789, 100M, 1, 568, 100M, 0, False, CTGTGTCTTTTTTCTTTTATTGTCATTGCCTGGTGTGAGCCCTGTTTTTATTTCCGGGGCTGCAATGCCAAAGACAGCATTCGGCTTGGACGTTTTGGTT
r268288|4|1|1|512873|100=|0|512723|100=, 3, 4, 512723, 100M, 1, 634, 100M, 0, False, GCCAAAGACAGCATTCGGCTTGGACGTTTTGGTTTATCTGTTGGCTTGCACCCAGCACGCTTTGGGGCCCCGCCTGGAGAGCCCAGACTCCTCCGAGCAC
r229595|4|0|0|512593|100=|1|512709|100=, 3, 4, 512709, 100M, 1, 648, 100M, 0, False, TCGGCTTGGACGTTTTGGTTTATCTGTTGGCTTGCACCCAGCACGCTTTGGGGCCCCGCCTGGAGAGCCCAGACTCCTCCGAGCACCCCCATGCTTTACG
r229595|4|0|0|512593|100=|1|512709|100=, 3, 4, 512593, 100M, 1, 764, 100M, 0, False, CTTCCTATGACAATTCGTTATCTGGAACTGGCGATCTCTTTGTTCGGTGCGTCCGTACACTTTTGTCGTTGCATGTTTTTGCGCCCAACATTTGGTGTTC
r85872|2|0|0|455834|100=|1|456006|100=, 3, 2, 456006, 100M, 1, 902, 100M, 0, False, ATCAAGCATGTCGTACGTTCCGCGGCGTATATTCAGTATAGAACGTCTTCTCATATTCATCATGGACTCCAACGCATTGCGCTCTCTTTAATGTTTCTGC
</pre></div>
</div>
<p>We all know how exciting it is to scroll through a .csv file, so we have some additional tools to analyse the misaligned
reads and display the information graphically (These require <a class="reference external" href="http://matplotlib.org/index.html">Matplotlib</a> to be installed).</p>
<p><em>(For aesthetic reasons, we will use a different reads file that has reads from all the chromosomes, but at lower coverage
so it doesn&#8217;t take so long to run)</em></p>
<div class="highlight-text"><div class="highlight"><pre>$ ../examples/demo/low_coverage_reads.sh
...
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ plot_align circle ../examples/demo/Out/null_reads_low  --down-sample 2
</pre></div>
</div>
<img alt="_images/null_reads_low_circle_plot.png" src="_images/null_reads_low_circle_plot.png" />
<p>This is a plot of where the reads end up after alignment. As can be seen, while a small percentage of reads end up in
different chromosomes, most of the misalignments are local (the lines end up on the same chromosome).</p>
<div class="highlight-text"><div class="highlight"><pre>$ plot_align matrix ../examples/demo/Out/null_reads_low
</pre></div>
</div>
<img alt="_images/null_reads_low_matrix_plot.png" src="_images/null_reads_low_matrix_plot.png" />
</div>
</div>
<div class="section" id="taking-reads-just-from-regions-around-variants">
<h2>2.4. Taking reads just from regions around variants<a class="headerlink" href="#taking-reads-just-from-regions-around-variants" title="Permalink to this headline">¶</a></h2>
<p>By setting the <code class="docutils literal"><span class="pre">variants_only</span></code> flag to <em>True</em> in the parameters file we can selectively generate reads from regions bordering
a variation.</p>
<div class="highlight-json"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>13
14
15
16
17</pre></div></td><td class="code"><div class="highlight"><pre>  },
  &quot;chromosomes&quot;: [1, 2, 3, 4],
  &quot;variants_only&quot;: true,
  &quot;read_model&quot;: &quot;simple_illumina&quot;,
  &quot;coverage&quot;: 30,
</pre></div>
</td></tr></table></div>
<div class="highlight-text"><div class="highlight"><pre>$ reads generate  ../examples/demo/illumina_reads_variants_only.json -v
DEBUG:mitty.reads:Took 9.440709s to write 31816 templates
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ bwa mem -t 8 -p  ../examples/data/red_alga.fa.gz  ../examples/demo/Out/vreads_c.fq &gt; ../examples/demo/Out/temp.sam
[M::main_mem] read 63632 sequences (6363200 bp)...
...
[main] CMD: bwa mem -t 8 -p ../examples/data/red_alga.fa.gz ../examples/demo/Out/vreads_c.fq
[main] Real time: 2.530 sec; CPU: 7.407 sec
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ samtools view -Sb  ../examples/demo/Out/temp.sam &gt; ../examples/demo/Out/temp.bam
[samopen] SAM header is present: 20 sequences.
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ samtools sort  ../examples/demo/Out/temp.bam  ../examples/demo/Out/vreads
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre>$ samtools index  ../examples/demo/Out/vreads.bam
</pre></div>
</div>
<p>We can use an alignment browser, such as Tablet, to see the alignments</p>
<img alt="_static/vreads_bam_tablet.png" src="_static/vreads_bam_tablet.png" />
<p>Note how the reads are restricted to the variant locations, as we asked for.</p>
</div>
<div class="section" id="analysing-variant-callers">
<h2>2.5. Analysing variant callers<a class="headerlink" href="#analysing-variant-callers" title="Permalink to this headline">¶</a></h2>
<p>TODO</p>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="additional_utils.html" class="btn btn-neutral float-right" title="3. Additional Utilities">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="introduction.html" class="btn btn-neutral" title="1. Overview"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014,Seven Bridges Genomics
(kaushik.ghose@sbgenomics.com).
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.1dev',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>