

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>6. Ramblings AKA. Diary of a Developer &mdash; Mitty 1.1.1dev documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Mitty 1.1.1dev documentation" href="index.html"/>
        <link rel="prev" title="5. Stock Models" href="models.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="index.html" class="fa fa-home"> Mitty</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">1. Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#installation">1.1. Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#trivia">1.2. Trivia</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">2. Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#creating-genomes">2.1. Creating genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#taking-reads">2.2. Taking reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#testing-alignment-accuracy-of-bwa-mem">2.3. Testing alignment accuracy of BWA MEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#taking-reads-just-from-regions-around-variants">2.4. Taking reads just from regions around variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#analysing-variant-callers">2.5. Analysing variant callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="additional_utils.html">3. Additional Utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="additional_utils.html#plot-gc-bias-plot-gc-bias-of-a-bam-file">3.1. <cite>plot_gc_bias</cite>: Plot GC bias of a BAM file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="theory_and_algorithms.html">4. Theory and algorithms</a><ul>
<li class="toctree-l2"><a class="reference internal" href="theory_and_algorithms.html#population-simulations">4.1. Population simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Stock Models</a><ul>
<li class="toctree-l2"><a class="reference internal" href="models.html#variant-models">5.1. Variant models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#population-models">5.2. Population models</a></li>
<li class="toctree-l2"><a class="reference internal" href="models.html#read-models">5.3. Read models</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">6. Ramblings AKA. Diary of a Developer</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#simulating-a-million-genomes-in-python">6.1. Simulating a million Genomes in Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="#report-generation">6.2. Report generation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Mitty</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>6. Ramblings AKA. Diary of a Developer</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/ramblings.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="ramblings-aka-diary-of-a-developer">
<h1>6. Ramblings AKA. Diary of a Developer<a class="headerlink" href="#ramblings-aka-diary-of-a-developer" title="Permalink to this headline">¶</a></h1>
<p>(In which I write down some of my design decisions.)</p>
<div class="section" id="simulating-a-million-genomes-in-python">
<h2>6.1. Simulating a million Genomes in Python<a class="headerlink" href="#simulating-a-million-genomes-in-python" title="Permalink to this headline">¶</a></h2>
<div class="section" id="memory-considerations">
<h3>6.1.1. Memory considerations<a class="headerlink" href="#memory-considerations" title="Permalink to this headline">¶</a></h3>
<p>In the early days Python was a no-brainer in terms of prototyping ideas. It got us off the ground very fast, it allowed
me to write comprehensive and compact code and to use things like docopts and sphinx that allowed good command line
interfaces and painless, excellent documentation generation.</p>
<p>The real challenge started when we were thinking of population simulations. The primary consideration was that, being a
simulation meant for benchmarking and tool testing, the code should be accurate, rather than fast, and if we wanted to
simulate a million genomes, we should be able to do it i.e. we would sacrifice time for space.</p>
<p>First, some numbers:</p>
<p>A single variant requires at least 7 bytes.</p>
<p>pos      4 bytes
ref      1 byte
alt      1 byte
genotype 1 byte</p>
<p>For speed of computation we like to keep a stop field which says where the variant stops (mostly pos + len(ref)) bringing
us to 11 bytes</p>
<p>pos      4 bytes
stop     4 bytes
ref      1 byte
alt      1 byte
genotype 1 byte</p>
<p>A single genome has about 3e6 (three million) variants (you see we are only considering snps, but this gets us a good
estimate of magnitude). This means that a single genome requires between 21e6 and 33e6 bytes.</p>
<p>The 1000G data set has about 100e6 (hundred million) unique variants from 2500 samples, and a million genome data will
probably have a billion.</p>
<p>Considering a population of only 2500 we see that</p>
<p>If we store each genome separately, we will require between 52.5 and 82.5 billion bytes. If we store just the unique
variants we will end up requiring 0.7 to 1.1 billion bytes + something for the genotypes of each sample.</p>
<p>How would the second scheme work?</p>
<p>We would use 6 or 10 bytes to store just the variant (no genotype information) and then for each sample just store the
genotype information and a pointer to the variant, which takes the form:</p>
<p>index    4 bytes  (points to the variant in the master list)
genotype 1 byte</p>
<p>(In practice, I compressed this down to 4 bytes, by using 30 bits for the index and 2 bits for the diploid genotype)</p>
<p>0.6 to 1.0 billion bytes for the master list of variants.
15 million bytes per sample = 37.5 billion bytes for 2500 samples.</p>
<p>So it is clear that even with a &#8220;tiny&#8221; sample like the 1000G data we can not fit the whole population in memory at the
same time, but by using a master list of unique variants we can halve our memory requirements.</p>
</div>
<div class="section" id="sampling-a-population-into-existence">
<h3>6.1.2. Sampling a population into existence<a class="headerlink" href="#sampling-a-population-into-existence" title="Permalink to this headline">¶</a></h3>
<p>While I was</p>
</div>
</div>
<div class="section" id="report-generation">
<h2>6.2. Report generation<a class="headerlink" href="#report-generation" title="Permalink to this headline">¶</a></h2>
<p>Static textual html pages are pretty easy to generate with pure Python. HTML pages with static or interactive graphs are a
little harder but some distance can be covered using <a class="reference external" href="http://matplotlib.com">Matplotlib</a> and the <a class="reference external" href="http://mpld3.com">mpld3</a> library. If we are willing to abandon
Matplotlib, we can use <a class="reference external" href="http://bokeh.io">Bokeh</a> which is perfect for this. Matplotlib, however, has a high bar for publication type static
figures that are written to pdf.</p>
<p>The challenge comes when we
need interactive plots that show large datasets. Loading all the data in at once causes the page to become sluiggish.</p>
<p>In a pure Matplotlib interactive figure we can use callbacks to load views of the data from disk as needed. There is no
way to translate this to a web-page, currently.</p>
<p>Using Bokeh we can build such a plot, but we need a bokeh server to serve the page and
respond to the interaction commands. This is not feasible with our current platform.</p>
<p>From these considerations I decided to use the following approach for the platform:</p>
<ol class="arabic simple">
<li>There is a single page html report generated that can be easily seen on the platform</li>
<li>The html has some interactive textual tables summarizing benchmark data.</li>
</ol>
<p>3. There are a few interactive summary plots which attempt to give a graphical overview of the errors. The summary plots
bin the data so that we handle a small quantity of numbers which keep our web-page snappy.</p>
<p>For people using the terminal, we generate interactive plots in Bokeh/Matplotlib that load data from disk as needed to
display detailed views of the errors for extensive debugging.</p>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="models.html" class="btn btn-neutral" title="5. Stock Models"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014-2015,Seven Bridges Genomics
(kaushik.ghose@sbgenomics.com).
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.1dev',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>